<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<title>gtscript documentation</title>
<link rel="stylesheet" type="text/css" href="style.css">
</head>
<body>
<div id="menu">
<ul>
<li><a href="index.html">Overview</a></li>
<li><a href="pub/">Download</a></li>
<li><a href="cgi-bin/gitweb.cgi?p=genometools.git;a=summary">Browse source</a></li>
<li><a href="design.html">Design</a></li>
<li><a id="current" href="docs.html"><tt>gtscript</tt> docs</a></li>
<li><a href="libgtview.html"><tt>libgtview</tt> library</a></li>
<li><a href="license.html">License</a></li>
</ul>
</div>
<div id="main">
<h1><tt>gtscript</tt> documentation</h1>
<p>
This document describes the <tt>gtscript</tt> API. <tt>gtscript</tt> is
basically <a href="http://www.lua.org/">Lua</a> (an embeddable scripting
language) plus parts of the <i>GenomeTools</i> C libraries exported to
Lua. Because the <i>GenomeTools</i> binary <tt>gt</tt> contains an embedded Lua
interpreter, <tt>gtscript</tt> files can be executed with the <tt>gt</tt>
binary.
The parts of the <i>GenomeTools</i> C libraries exported to Lua (the
<tt>gtscript</tt> API) are described here, for a documentation of Lua itself and
its APIs, please refer to the <a href="http://www.lua.org/manual/5.1/">Lua
reference manual</a>.
</p>
<h2>Notes</h2>
<ul>
<li>
You have to add <code>require 'gtlua'</code> to your script
in order to load the parts of the <tt>gtscript</tt> API which are implemented in
<tt>gtscript</tt> itself.
<li>
By default, all functions of the <tt>gtscript</tt> API are contained in the
<code>gt</code> table.  That is, you have to prepend all functions calls with
``<code>gt.</code>''. For example, write <code>gt.show()</code> to call the
<code>show()</code> function documented below.  You can use
<code>gt.export()</code> or <code>gt.re()</code> to export the <code>gt</code>
table to the global environment, which makes the prepending unnecessary (but
clutters your global environment).
<li>
If a function is documented as ``returns an array'', this means that the function
returns a table where only consecutive
integer keys from 1 on are used and which can be traversed with the
<a
href="http://www.lua.org/manual/5.1/manual.html#pdf-ipairs"><code>ipairs()</code></a> function.
</ul>
<h2>Classes</h2>
<ul>

  <li><a href="#Alpha">Alpha</a>

  <li><a href="#Bittab">Bittab</a>

  <li><a href="#CDSStream">CDSStream</a>

  <li><a href="#CSAStream">CSAStream</a>

  <li><a href="#Diagram">Diagram</a>

  <li><a href="#FeatureIndex">FeatureIndex</a>

  <li><a href="#FeatureStream">FeatureStream</a>

  <li><a href="#FeatureVisitor">FeatureVisitor</a>

  <li><a href="#GenomeNode">GenomeNode</a>

  <li><a href="#GenomeNodeIterator">GenomeNodeIterator</a>

  <li><a href="#GenomeStream">GenomeStream</a>

  <li><a href="#Range">Range</a>

  <li><a href="#RegionMapping">RegionMapping</a>

  <li><a href="#Render">Render</a>

  <li><a href="#ScoreMatrix">ScoreMatrix</a>

  <li><a href="#StreamEvaluator">StreamEvaluator</a>

</ul>
<h2>Sole functions</h2>
<code>genome_feature_new(type, range, strand)</code>
<p>
Returns a new genome feature of <code>type</code> spanning <code>range</code> on <code>strand</code>.
</p>
<hr>
<code>sequence_region_new(seqid, range)</code>
<p>
Returns a new sequence region for sequence id <code>seqid</code> spanning <code>range</code>.
</p>
<hr>
<code>genome_feature:get_strand()</code>
<p>
Returns the strand of <code>genome_feature</code>.
</p>
<hr>
<code>genome_feature:get_source()</code>
<p>
Returns the source of <code>genome_feature</code>.
</p>
<hr>
<code>genome_feature:set_source(source)</code>
<p>
Set the source of <code>genome_feature</code> to <code>source</code>.
</p>
<hr>
<code>genome_feature:output_leading()</code>
<p>
Show leading part of GFF3 output for <code>genome_feature</code>
</p>
<hr>
<code>genome_feature:get_type()</code>
<p>
Return type of <code>genome_feature</code> as string.
</p>
<hr>
<code>genome_feature:extract_sequence(type, join, region_mapping)</code>
<p>
Extract the sequence of <code>genome_feature</code>. If <code>join</code> is false and <code>genome_feature</code> has type <code>type</code> the sequence is returned (using <code>region_mapping</code> to get it). If <code>join</code> is true and <code>genome_feature</code> has children of type <code>type</code> their joined sequences are returned. If none of the above applies nil is returned.
</p>
<hr>
<code>gff3_in_stream_new_sorted(filename)</code>
<p>
Returns a new GFF3 input stream object for <code>filename</code>. The file <code>filename</code> has to be a sorted GFF3 file.
</p>
<hr>
<code>gff3_out_stream_new(genome_stream)</code>
<p>
Returns a new GFF3 output stream which pulls its features from <code>genome_stream</code>.
</p>
<hr>
<code>gff3_visitor_new()</code>
<p>
Returns a new GFF3 visitor.
</p>
<hr>
<code>ranges_sort(range_array)</code>
<p>
Returns an array containing the ranges from array <code>range_array</code> in sorted order.
</p>
<hr>
<code>ranges_are_sorted(range_array)</code>
<p>
Returns true if the ranges in array <code>range_array</code> are sorted, false otherwise.
</p>
<hr>
<code>translate_dna(dna)</code>
<p>
Returns translated <code>dna</code>.
</p>
<hr>
<code>reload()</code>
<p>
Reload <code>gt</code> module.
</p>
<hr>
<code>features_contain_marked(features)</code>
<p>
Returns true if the given array of <code>features</code> contains a marked feature, false otherwise.
</p>
<hr>
<code>features_show(features)</code>
<p>
Print the given array of <code>features</code> to stdout.
</p>
<hr>
<code>features_get_marked(features)</code>
<p>
Return all marked <code>features</code> (an array) as an array or nil if <code>features</code> contains no marked features.
</p>
<hr>
<code>features_show_marked(features)</code>
<p>
Print all marked <code>features</code> (an array) to stdout.
</p>
<hr>
<code>features_mRNAs2genes(in_features)</code>
<p>
Return an array of genome features which contains a separate gene feature for each mRNA in <code>in_features</code>.
</p>
<hr>
<code>features_extract_sequences(features, type, join, region_mapping)</code>
<p>
Return an array with the sequences of the given features.
</p>
<hr>
<code>export()</code>
<p>
Export the content of <code>gt</code> table to the global environment.
</p>
<hr>
<code>display(filename)</code>
<p>
Call external 'display' program for file <code>filename</code>.
</p>
<hr>
<code>show_table(tbl)</code>
<p>
Show all keys and values of table <code>tbl</code>.
</p>
<hr>
<code>show(all)</code>
<p>
Show content of the <code>gt</code> table.
</p>
<hr>
<code>re()</code>
<p>
Reload the <code>gt</code> module and export its content to the global environment.
</p>
<hr>
<a name="Alpha"></a>
<h2>Class Alpha</h2>
<code>alpha_new_protein()</code>
<p>
Returns a new protein alphabet.
</p>
<hr>
<code>alpha:decode(code)</code>
<p>
Returns a string containing the decoded character of the <code>code</code> number.
</p>
<hr>
<code>alpha:size()</code>
<p>
Returns the size of <code>alpha</code> a number.
</p>
<hr>
<a name="Bittab"></a>
<h2>Class Bittab</h2>
<code>bittab_new(num_of_bits)</code>
<p>
Returns a bittab with <code>num_of_bits</code> many bits.
</p>
<hr>
<code>bittab:set_bit(bit)</code>
<p>
Set <code>bit</code> in <code>bittab</code>.
</p>
<hr>
<code>bittab:unset_bit(bit)</code>
<p>
Unset <code>bit</code> in <code>bittab</code>.
</p>
<hr>
<code>bittab:complement(src)</code>
<p>
Store the complement of bittab <code>src</code> in <code>bittab</code>. <code>bittab</code> and <code>src</code> must have the same size.
</p>
<hr>
<code>bittab:equal(src)</code>
<p>
Set <code>bittab</code> equal to bittab <code>src</code>. <code>bittab</code> and <code>src</code> must have the same size.
</p>
<hr>
<code>bittab:and_equal(src)</code>
<p>
Set <code>bittab</code> equal to the bitwise AND of <code>bittab</code> and <code>src</code>. <code>bittab</code> and <code>src</code> must have the same size.
</p>
<hr>
<code>bittab:bit_is_set(bit)</code>
<p>
Returns true if <code>bit</code> is set in <code>bittab</code>, false otherwise.
</p>
<hr>
<a name="CDSStream"></a>
<h2>Class CDSStream</h2>
<code>cds_stream_new(region_mapping)</code>
<p>
Returns a new CDS (coding sequence) stream object (a genome stream) which uses genome stream <code>in_stream</code> as input. The CDS stream adds CDS features to exon features in <code>in_stream</code>. The given <code>region_mapping</code> is used to map the sequence regions given in <code>in_stream</code> to the actual sequence files necessary for computing the coding sequences.
</p>
<hr>
<a name="CSAStream"></a>
<h2>Class CSAStream</h2>
<code>csa_stream_new(in_stream, join)</code>
<p>
Returns a new CSA (consensus spliced alignment) stream object (a genome stream) which uses genome stream <code>in_stream</code> as input. The CSA stream replaces spliced alignments with computed consensus spliced alignments. The optional <code>join</code> parameters sets the length for the spliced alignment clustering (default: 300).
</p>
<hr>
<a name="Diagram"></a>
<h2>Class Diagram</h2>
<code>diagram_new(feature_index, range, seqid)</code>
<p>
Return a diagram object which contains the genome nodes given in <code>feature_index</code> in the given <code>range</code> of the sequence region with sequence ID <code>seqid</code>.
</p>
<hr>
<a name="FeatureIndex"></a>
<h2>Class FeatureIndex</h2>
<code>feature_index_new()</code>
<p>
Returns a new <code>feature_index</code> object.
</p>
<hr>
<code>feature_index:add_sequence_region(sequence_region)</code>
<p>
Add <code>sequence_region</code> to <code>feature_index</code>.
</p>
<hr>
<code>feature_index:add_genome_feature(genome_feature)</code>
<p>
Add <code>genome_feature</code> to <code>feature_index</code>.
</p>
<hr>
<code>feature_index:get_features_for_seqid(seqid)</code>
<p>
Returns the genome features for sequence ID <code>seqid</code> in an array.
</p>
<hr>
<code>feature_index:get_features_for_range(seqid, range)</code>
<p>
Returns the genome features for sequence ID <code>seqid</code> within <code>range</code> in an array.
</p>
<hr>
<code>feature_index:get_first_seqid()</code>
<p>
Returns the first sequence ID stored in <code>feature_index</code>.
</p>
<hr>
<code>feature_index:get_seqids()</code>
<p>
Returns an array containins all sequence IDs stored in <code>feature_index</code>.
</p>
<hr>
<code>feature_index:get_range_for_seqid(seqid)</code>
<p>
Returns the range covered by features of sequence ID <code>seqid</code> in <code>feature_index</code>.
</p>
<hr>
<code>feature_index:get_coverage(seqid, maxdist)</code>
<p>
Computes the coverage for the sequence ID <code>seqid</code>. The optional <code>maxdist</code> parameter denotes the maximal distance two features can be apart without creating a new Range. Returns an array of Ranges denoting parts the of <code>seqid</code> covered by features.
</p>
<hr>
<code>feature_index:get_marked_regions(seqid, maxdist)</code>
<p>
Returns an array of Ranges denoting parts of <code>seqid</code> which are covered by at least one marked feature. Internally, <code>get_coverage()</code> is called and the <code>maxdist</code> is passed along.
</p>
<hr>
<code>feature_index:render_to_png(seqid, range, png_file, width)</code>
<p>
Render to PNG file <code>png_file</code> for <code>seqid</code> in <code>range</code> with optional <code>width</code>. If no <code>png_file</code> is given <code>os.tmpname()</code> is called to create one. Returns name of written PNG file.
</p>
<hr>
<code>feature_index:show_seqids()</code>
<p>
Show all sequence IDs.
</p>
<hr>
<code>feature_index:get_all_features()</code>
<p>
Returns all features from <code>feature_index</code>.
</p>
<hr>
<a name="FeatureStream"></a>
<h2>Class FeatureStream</h2>
<code>feature_stream_new(in_stream, feature_index)</code>
<p>
Returns a new feature stream object (a genome stream) over <code>feature_index</code> which uses the genome stream <code>in_stream</code> as input. That is, all genome nodes which are pulled through the feature stream are added to the <code>feature_index</code>.
</p>
<hr>
<a name="FeatureVisitor"></a>
<h2>Class FeatureVisitor</h2>
<code>feature_visitor_new(feature_index)</code>
<p>
Returns a new feature visitor object over <code>feature_index</code>. That is, all genome nodes which are visited by the feature visitor are added to the <code>feature_index</code>.
</p>
<hr>
<a name="GenomeNode"></a>
<h2>Class GenomeNode</h2>
<code>genome_node:get_filename()</code>
<p>
Returns the filenname of <code>genome_node</code>.
</p>
<hr>
<code>genome_node:get_range()</code>
<p>
Returns the range of <code>genome_node</code>.
</p>
<hr>
<code>genome_node:get_seqid()</code>
<p>
Returns the sequence id of <code>genome_node</code>.
</p>
<hr>
<code>genome_node:set_seqid(seqid)</code>
<p>
Set the sequence id of <code>genome_node</code> to <code>seqid</code>.
</p>
<hr>
<code>genome_node:accept(genome_visitor)</code>
<p>
Accept <code>genome_visitor</code>.
</p>
<hr>
<code>genome_node:is_part_of_genome_node(child_node)</code>
<p>
Make <code>genome_node</code> the parent of <code>child_node</code>.
</p>
<hr>
<code>genome_node:mark()</code>
<p>
Mark <code>genome_node</code>.
</p>
<hr>
<code>genome_node:is_marked()</code>
<p>
Returns true if <code>genome_node</code> is marked, false otherwise.
</p>
<hr>
<code>genome_node:contains_marked()</code>
<p>
Returns true if <code>genome_node</code> contains a marked node, false otherwise.
</p>
<hr>
<code>genome_node:show(gff3_visitor)</code>
<p>
Show genome node on stdout (using the optional <code>gff3_visitor</code>).
</p>
<hr>
<code>genome_node:show_marked()</code>
<p>
Show marked parts of genome node on stdout.
</p>
<hr>
<a name="GenomeNodeIterator"></a>
<h2>Class GenomeNodeIterator</h2>
<code>genome_node_iterator_new(genome_node)</code>
<p>
Returns a new genome node iterator which performs a depth-first traversel of <code>genome_node</code> (including <code>genome_node</code> itself).
</p>
<hr>
<code>genome_node_iterator_new_direct(genome_node)</code>
<p>
Returns a noew genome node iterator wich iterates over all direct children of <code>genome_node</code> (without <code>genome_node</code> itself).
</p>
<hr>
<code>genome_node_iterator:next()</code>
<p>
Returns the next genome node for <code>genome_node_iterator</code> or nil.
</p>
<hr>
<a name="GenomeStream"></a>
<h2>Class GenomeStream</h2>
<code>genome_stream:next_tree()</code>
<p>
Returns the next genome node for <code>genome_stream</code> or nil.
</p>
<hr>
<a name="Range"></a>
<h2>Class Range</h2>
<code>range_new(startpos, endpos)</code>
<p>
Returns a new range object with start <code>startpos</code> and end <code>endpos</code>. <code>startpos</code> must be smaller or equal than <code>endpos</code>.
</p>
<hr>
<code>range:get_start()</code>
<p>
Returns start of <code>range</code>.
</p>
<hr>
<code>range:get_end()</code>
<p>
Returns end of <code>range</code>.
</p>
<hr>
<code>range:overlap(other_range)</code>
<p>
Returns true if <code>range</code> and <code>other_range</code> overlap, false otherwise.
</p>
<hr>
<code>range:show()</code>
<p>
Show range on stdout.
</p>
<hr>
<a name="RegionMapping"></a>
<h2>Class RegionMapping</h2>
<code>region_mapping_new_seqfile(seqfile)</code>
<p>
Returns a new region mapping which maps everything onto sequence file <code>seqfile</code>.
</p>
<hr>
<a name="Render"></a>
<h2>Class Render</h2>
<code>render_new()</code>
<p>
Returns a new render object.
</p>
<hr>
<code>render:to_png(diagram, filename, width)</code>
<p>
Uses <code>render</code> to render the given <code>diagram</code> as PNG to <code>filename</code>. The optional <code>width</code> parameter sets the width of the PNG (default: 800).
</p>
<hr>
<a name="ScoreMatrix"></a>
<h2>Class ScoreMatrix</h2>
<code>score_matrix_new_read_protein(path)</code>
<p>
Returns a new protein score matrix object which has been read from file <code>path</code>.
</p>
<hr>
<code>score_matrix:get_dimension()</code>
<p>
Returns the dimension of the <code>score_matrix</code> as number.
</p>
<hr>
<code>score_matrix:get_score(idx1, idx2)</code>
<p>
Returns the score for <code>idx1</code>,<code>idx2</code> as number.
</p>
<hr>
<a name="StreamEvaluator"></a>
<h2>Class StreamEvaluator</h2>
<code>stream_evaluator_new(reality_stream, prediction_stream)</code>
<p>
Returns a new stream evaluator object for the two genome streams <code>reality_stream</code> and <code>prediction_stream</code>.
</p>
<hr>
<code>stream_evaluator:evaluate(genome_visitor)</code>
<p>
Run evaluation of <code>stream_evaluator</code>. All evaluated features are visited by the optional <code>genome_visitor</code>.
</p>
<hr>
<code>stream_evaluator:show()</code>
<p>
Show result of <code>stream_evaluator</code> on stdout.
</p>
<hr>
<div id="footer">
Copyright &copy; 2008
<a href="mailto:gremme@gmail.com">Gordon Gremme.</a>
Last update: 2008-02-28
</div>
</div>
<script src="http://www.google-analytics.com/urchin.js" type="text/javascript">
</script>
<script type="text/javascript">
_uacct = "UA-301256-4";
urchinTracker();
</script>
</body>
</html>
